// @adalo/cli/src/login.js
const prompt = require('prompt')
const fs = require('fs')
const path = require('path')
const os = require('os')
const chalk = require('chalk')
const axios = require('axios')

const server =
  process.env.ADALO_BACKEND_SERVER ||
  process.env.FOUNDRY_BACKEND_SERVER ||
  'https://adalo-backend.herokuapp.com'

const schema = {
  properties: {
    email: {
      required: true,
    },
    password: {
      required: true,
      hidden: true,
    },
  },
}

const login = (args, opts) => {
  let server = 'https://backend.adalo.com'

  if (opts.local) {
    server =
      process.env.ADALO_BACKEND_SERVER ||
      process.env.FOUNDRY_BACKEND_SERVER ||
      'http://localhost:8084'
  }

  prompt.start()

  prompt.get(schema, async (err, result) => {
    if (err) {
      console.log('ERROR:', err)
      return
    }

    try {
      const { data } = await axios.post(`${server}/sessions`, result)
      const { success, sessionToken } = data

      if (success) {
        const rc = path.join(os.homedir(), '.adalorc')

        fs.writeFileSync(
          rc,
          JSON.stringify({
            authToken: sessionToken,
          })
        )

        console.log(`\n  Login successful!`.bold.green)
        console.log(
          `  Your login information was saved to`,
          `${rc}\n`.italic.bold.green
        )
      }
    } catch (error) {
      console.error('')
      console.error('  Login failed.'.bold.red)
      console.error('  Please try again.')
      console.error('')
      console.error(
        chalk.bold(
          `  For further documentation please visit ${chalk.italic.blue(
            'https://developers.adalo.com'
          )}`
        )
      )
      console.error('')
      process.exit(1)
    }
  })
}

module.exports = login
